<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Psichomics by nuno-agostinho</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Psichomics</h1>
        <p>R package with a visual interface to quantify, analyse and visualise alternative splicing data</p>

        <p class="view"><a href="https://github.com/nuno-agostinho/psichomics">View the Project on GitHub <small>nuno-agostinho/psichomics</small></a></p>


        <ul>
          <li><a href="https://github.com/nuno-agostinho/psichomics/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/nuno-agostinho/psichomics/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/nuno-agostinho/psichomics">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <p><a href="https://travis-ci.org/nuno-agostinho/psichomics"><img src="https://travis-ci.org/nuno-agostinho/psichomics.svg?branch=master" alt="Build Status"></a> <a href="https://ci.appveyor.com/project/nuno-agostinho/psichomics"><img src="https://ci.appveyor.com/api/projects/status/github/nuno-agostinho/psichomics?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a> <a href="https://codecov.io/gh/nuno-agostinho/psichomics"><img src="https://codecov.io/gh/nuno-agostinho/psichomics/branch/master/graph/badge.svg" alt="codecov"></a></p>

<h2>
<a id="table-of-contents" class="anchor" href="#table-of-contents" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Table of Contents</h2>

<ul>
<li>
<a href="#install-and-start-running">Install and Start Running</a>

<ul>
<li><a href="#development-version">Development Version</a></li>
</ul>
</li>
<li><a href="#tutorials">Tutorials</a></li>
<li>
<a href="#data-input">Data Input</a>

<ul>
<li><a href="#download-tcga-data">Download TCGA data</a></li>
<li><a href="#load-user-files">Load User Files</a></li>
</ul>
</li>
<li><a href="#splicing-quantification">Splicing Quantification</a></li>
<li>
<a href="#data-analyses">Data Analyses</a>

<ul>
<li><a href="#differential-splicing-analysis">Differential Splicing Analysis</a></li>
<li><a href="#principal-component-analysis-pca">Principal Component Analysis (PCA)</a></li>
<li><a href="#survival-analysis">Survival Analysis</a></li>
<li><a href="#gene-transcript-and-protein-information">Gene, Transcript and Protein Information</a></li>
</ul>
</li>
<li><a href="#data-groups">Data Groups</a></li>
<li><a href="#feedback">Feedback</a></li>
<li><a href="#contributions">Contributions</a></li>
<li><a href="#references">References</a></li>
</ul>

<h2>
<a id="install-and-start-running" class="anchor" href="#install-and-start-running" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install and Start Running</h2>



<h3>
<a id="development-version" class="anchor" href="#development-version" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Development Version</h3>

<p>To install and start using the development version, follow these steps:</p>

<ol>
<li><a href="https://www.r-project.org">Install R</a></li>
<li>Depending on your operative system, install:

<ul>
<li>
<a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a> (Windows)</li>
<li>
<a href="https://developer.apple.com/downloads">Xcode command-line tools</a> (Mac)</li>
<li>
<strong>r-devel</strong> or <strong>r-base-dev</strong> (Linux)</li>
</ul>
</li>
<li>Open <a href="https://www.rstudio.com/products/rstudio">RStudio</a> (or open a console, type <code>R</code> and press enter)</li>
<li>Type the following to install Bioconductor: 

<ul>
<li><code>source("https://bioconductor.org/biocLite.R"); biocLite("BiocInstaller")</code></li>
</ul>
</li>
<li>Type the following to install, load and start the visual interface:</li>
</ol>

<div class="highlight highlight-source-r"><pre>install.packages(<span class="pl-s"><span class="pl-pds">"</span>devtools<span class="pl-pds">"</span></span>)
<span class="pl-e">devtools</span><span class="pl-k">::</span>install_github(<span class="pl-s"><span class="pl-pds">"</span>nuno-agostinho/psichomics<span class="pl-pds">"</span></span>)
library(<span class="pl-smi">psichomics</span>)
psichomics()</pre></div>

<h2>
<a id="tutorials" class="anchor" href="#tutorials" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Tutorials</h2>

<p>The following tutorials are available:</p>

<ul>
<li><a href="http://rpubs.com/nuno-agostinho/psichomics-tutorial-visual">Visual interface</a></li>
<li><a href="http://rpubs.com/nuno-agostinho/psichomics-cli-tutorial">Command-line interface</a></li>
</ul>

<p>Other tutorials coming soon:</p>

<ul>
<li>Developers and other contributors</li>
</ul>

<h2>
<a id="data-input" class="anchor" href="#data-input" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data Input</h2>

<h3>
<a id="download-tcga-data" class="anchor" href="#download-tcga-data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download TCGA Data</h3>

<p>You can download data from <a href="https://tcga-data.nci.nih.gov">The Cancer Genome Atlas (TCGA)</a> using this
package. Simply choose the cohort of interest, date of the sample, type of 
interest and so on.</p>

<h3>
<a id="load-user-files" class="anchor" href="#load-user-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Load User Files</h3>

<p>To load your own files, simply choose the folder where the data is located. 
PSIchomics will try to process all the data contained in the given folder and
sub-folders to search for files that can be loaded.</p>

<h2>
<a id="splicing-quantification" class="anchor" href="#splicing-quantification" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Splicing Quantification</h2>

<p>The quantification of each alternative splicing event is based on the proportion
of junction reads that support the inclusion isoform, known as percent 
spliced-in or PSI <a href="#references">(Wang <em>et al.</em>, 2008)</a>.</p>

<p>An estimate of this value is obtained based on the the proportion of reads 
supporting the inclusion of an exon over the reads supporting both the inclusion
and exclusion of that exon. To measure this estimate, both alternative splicing 
annotation and the quantification of RNA-Seq reads aligning to splice junctions
(junction quantification) are required. While alternative splicing Human (hg19
assembly) annotation is provided by the package, junction quantification may be
retrieved from <a href="https://tcga-data.nci.nih.gov">TCGA</a>.</p>

<h2>
<a id="data-analyses" class="anchor" href="#data-analyses" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data Analyses</h2>

<p>The program performs survival and principal component analyses, as well as
differential splicing analysis using non-parametric statistical tests.</p>

<h3>
<a id="differential-splicing-analysis" class="anchor" href="#differential-splicing-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Differential Splicing Analysis</h3>

<p>Analyse alternative splicing quantification based on variance and median 
statistical tests. The groups available for differential analysis comprise 
sample types (e.g. normal versus tumour) and clinical attributes of patients 
(e.g. tumour stage).</p>

<h3>
<a id="principal-component-analysis-pca" class="anchor" href="#principal-component-analysis-pca" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Principal Component Analysis (PCA)</h3>

<p>Explore alternative splicing quantification groups using associated clinical 
attributes.</p>

<h3>
<a id="survival-analysis" class="anchor" href="#survival-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Survival Analysis</h3>

<p>Analyse survival based on clinical attributes (e.g. tumour stage, gender and
race). Additionally, study the impact of the quantification of a single 
alternative splicing event on patient survivability.</p>

<h3>
<a id="gene-transcript-and-protein-information" class="anchor" href="#gene-transcript-and-protein-information" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Gene, Transcript and Protein Information</h3>

<p>For a given splicing event, examine its gene's annotation and corresponding 
transcripts and proteins. Related research articles are also available.</p>

<h2>
<a id="data-groups" class="anchor" href="#data-groups" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data Groups</h2>

<ul>
<li>
<strong>By column:</strong> automatically create groups by selecting a specific column of 
the dataset; for instance, to create a group for each tumour stage, start typing
<code>tumor_stage</code>, select the appropriate field from the suggestions, click 
<code>Create group</code> and confirm that there is now one group for each stage.</li>
<li>
<strong>By row:</strong> input specific rows to create a group</li>
<li>
<strong>By subset expression:</strong> type a subset expression</li>
<li>
<strong>By GREP expression:</strong> apply a GREP expression over a specific column of the 
dataset</li>
</ul>

<p>You can also select groups by clicking on them in order to merge, intersect or 
remove the groups.</p>

<h2>
<a id="feedback" class="anchor" href="#feedback" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Feedback</h2>

<p>All feedback on the program, documentation and associated material is welcome. 
Please, send any suggestions and comments to the following contact:</p>

<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunodanielagostinho@gmail.com">nunodanielagostinho@gmail.com</a>)</p>

<p><a href="http://imm.medicina.ulisboa.pt/group/compbio">Nuno Morais Lab, IMM</a></p>
</blockquote>

<h2>
<a id="contributions" class="anchor" href="#contributions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contributions</h2>

<p>Please note that this project is released with a <a href="CONDUCT.md">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms.</p>

<h2>
<a id="references" class="anchor" href="#references" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>References</h2>

<p>Wang, E. T., R. Sandberg, S. Luo, I. Khrebtukova, L. Zhang, C. Mayr, S. F. 
Kingsmore, G. P. Schroth, and C. B. Burge. 2008. <a href="http://www.nature.com/nature/journal/v456/n7221/full/nature07509.html"><em>Alternative isoform 
regulation in human tissue transcriptomes.</em></a> Nature 456 (7221): 470–76.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/nuno-agostinho">nuno-agostinho</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
